awscli/customizations/s3/utils.py

```diff
@@ -283,17 +286,45 @@ def find_chunksize(size, current_chunksize):
     the ``MAX_PARTS``.  If the ``chunksize`` is greater than
     ``MAX_SINGLE_UPLOAD_SIZE`` it returns ``MAX_SINGLE_UPLOAD_SIZE``.
     """
+    chunksize = enforce_max_parts(size, current_chunksize)
+    return enforce_chunksize_limits(chunksize)
+
+
+def enforce_chunksize_limits(current_chunksize):
     chunksize = current_chunksize
-    num_parts = int(math.ceil(size / float(chunksize)))
-    while num_parts > MAX_PARTS:
-        chunksize *= 2
-        num_parts = int(math.ceil(size / float(chunksize)))
+    chunksize_human = human_readable_size(chunksize)
     if chunksize > MAX_SINGLE_UPLOAD_SIZE:
+        LOGGER.debug(
+            "Chunksize greater than maximum chunksize. Setting to %s from %s."
+            % (human_readable_size(MAX_SINGLE_UPLOAD_SIZE), chunksize_human))
         return MAX_SINGLE_UPLOAD_SIZE
+    elif chunksize < MIN_UPLOAD_CHUNKSIZE:
+        LOGGER.debug(
+            "Chunksize less than minimum chunksize. Setting to %s from %s." %
+            (human_readable_size(MIN_UPLOAD_CHUNKSIZE), chunksize_human))
+        return MIN_UPLOAD_CHUNKSIZE
     else:
         return chunksize
 
 
+def enforce_max_parts(size, current_chunksize):
```

Comment:
Can we add docstrings for this.  One of things that would help is to make it clear that this is computing a new chunk size (and not really enforcing chunk size).

It would be helpful for this chunk adjustment functions to indicate whether they are guaranteed to increase/decrease the chunk size.  For example, this function will only ever return the current chunk size or a greater chunk size, never a smaller chunksize.